src/oled_display_node.cpp

```diff
@@ -242,6 +246,110 @@ int dispOled_writeBytes(dispCtx_t *dispCtx, uint8_t *outBuf, int numChars)
+      retCode = -2;
+      ROS_ERROR("Cannot open I2C def of %s with error %s", i2cDevFile, strerror(errno));
+      goto exitWithNoClose;
+    }
+
+    // The ioctl here will address the I2C slave device making it ready for 1 or more other bytes
+    if (ioctl(fd, I2C_SLAVE, address) != 0) {        // Set the port options and addr of the dev
+      retCode = -3;
+      ROS_ERROR("Failed to get bus access to I2C device %s!  ERROR: %s", i2cDevFile, strerror(errno));
+      goto exitWithFileClose;
+    }
+
+    if (chipRegAddr < 0) {     // Suppress reg address if negative value was used
+      buf[0] = (uint8_t)(chipRegAddr);          // Internal chip register address
+      if ((write(fd, buf, 1)) != 1) {           // Write both bytes to the i2c port
+        retCode = -4;
+        goto exitWithFileClose;
+      }
+    }
+
+    bytesRead = read(fd, pBuffer, numBytes);
+    if (bytesRead != numBytes) {      // verify the number of bytes we requested were read
+      retCode = -9;
+      goto exitWithFileClose;
+    }
+    retCode = bytesRead;
+
+  exitWithFileClose:
+    close(fd);
+
+  exitWithNoClose:
+
+  return retCode;
+}
+
+/*
+ * @name                dispOled_detectDisplayType
+ * @brief               Determine display type or if it is on the I2C bus
+ *
+ * @param               devName         Name of the I2C output device
+ * @param               i2c7bitAddr     7-bit I2C bus address
+ *
+ * @return              dispType        DISPLAY_TYPE_SSD1306,DISPLAY_TYPE_SH1106
+ * @return              Returns DISPLAY_TYPE_SH1106 [DEFAULT] or DISPLAY_TYPE_SSD1306
+ *                      Returns DISPLAY_TYPE_NONE for I2C error due to no device detected
+ */
+#define MX_DEV_NAME_LEN 32
+int dispOled_detectDisplayType(std::string devName, uint8_t i2c7bitAddr, int *dispType)
+{
+    uint8_t buf[16];
+    int     retCode = 0;
+    char    device[MX_DEV_NAME_LEN]; 
+    strncpy(&device[0], devName.c_str(), MX_DEV_NAME_LEN);
+    device[(MX_DEV_NAME_LEN-1)] = 0;     // protect against long dev names
+
+    // Read the status register at chip addr 0 to decide on chip type
+    int retCount = i2c_BufferRead(&device[0], i2c7bitAddr, &buf[0], 0, 1);
+    if (retCount < 0) {
+        ROS_ERROR("Error %d in reading OLED status register at 7bit I2CAddr 0x%x", 
+            retCount, i2c7bitAddr);
+        *dispType = DISPLAY_TYPE_NONE;
+        retCode = retCount;
+    } else if (retCount != 1) {
+        ROS_ERROR("Cannot read byte from OLED status register at 7bit Addr 0x%x", 
+            i2c7bitAddr);
+        *dispType = DISPLAY_TYPE_NONE;
+        retCode = -1;
+    } else {
+        ROS_INFO("Read OLED status register as 0x%02x", buf[0]);
+        if ((buf[0] & 0x07) == 0x06) {
+            // We found lower 3 bit as a 6 but datasheet does not spec it
+            *dispType = DISPLAY_TYPE_SSD1306;
+        } else {
+            // Data sheet guarentees lower 3 bits as 0
```

Comment:
I might be understanding wrongly but if the datasheet guarantees the lower 3 bits as 0 when the type is SH106, why wouldnt you rather check for that in the first if and then in the else assume that its a type SSD1306 if the if didnt pass? :)

Reply:
Although that sounds reasonable., what I was going for is our 'master default' is the SH1106.   So if it did not match SSD1306  then no matter WHAT it was it would use SH1106.    i am being concerned that it is best to default what production SHOULD have bought so only deviate from that if I get a certain match it is the SSD1306 thus test only for SSD1306.    Sadly there is no 'chipid' register like most every other I2C chip I have used AND I have used quite a few.